### ISSUE TITLE
Implement query-planner tests (i.e., Cradle) into this repository

### ISSUE THREAD

**Issue by abernix** [apollographql/rust#103]
With the landing of https://github.com/apollographql/apollo-server/pull/4297 in the `apollo-server` repository, we now have the "`.feature` files" which are the beginnings of what we are calling _Cradle_. 

These test files now act as compatibility acceptance tests for the current JavaScript query planner.  The `.feature` format is meant to be language-agnostic and introducing them to this repository should act as the basis for implementing the port of [our current `buildQueryPlan`](https://github.com/apollographql/apollo-server/blob/5a7fd856f6cbb1efe0ea91797cb81b9492bfe380/packages/apollo-gateway/src/buildQueryPlan.ts) into (I think) the `query-planner` module in this repository.

We imagine _Cradle_ growing into something more complete in the future (along with finding an agnostic home for them to match our intention for them to stand on their own), but that work will be tracked in another issue.  This issue is meant to track bringing those tests into this repository!

I believe the deliverables here are:

1. Be able to iteratively develop on implementation of a Rust `build_query_plan` using the `.feature` files introduced in https://github.com/apollographql/apollo-server/pull/4297.
2. Ensure that the same tests are also executed within CI.
3. All tests to be demonstrated as being executed, but failing, as there is no implementation yet! (And the implementation is out of scope for this issue)
4. Wholesale disabling of these tests in order to allow them to be incrementally disabled as the implementation grows to support their use-cases.  This disabling might just be demonstrated as a separate commit.

Just to repeat: the very large caveat here is that all the tests will be failing. 😄 

Some (possible!) considerations worth discussing, either along the way to or during exploration to implementing it might be:

* Determine _where_ the `.feature` files and the tests should live in this repository

  _Likely no need research extensively — as I think moves would be somewhat effortless — but some parity with best-practices or patterns of other repos utilizing Cucumber tests might be worth considering._

* Confirming that it's _okay_ to merely *copy* (yes, roughly, `cp there/*.feature here/.`) the `.feature` files into this repository.

  _**Copy**, you say?!  My claim would be that, while I DO think we need to find a more permanent home for them, that they are not currently changing at a rapid-enough pace that copying them will be a problematic ergonomic.  With this query-plan-porting-work in-progress, the JS query planner is effectively feature-frozen in its current state.  I think the source of truth could remain in the `apollo-server` repository until either the point the Rust query planner is shipped OR we answer the question: "where is the permanent home for Cradle tests?" and migrate them there.  (Also, I will open a separate issue for this!)_

* How to ignore `.feature` files.

  _I don't know enough about the implementation that will be necessary to understand what that would look like, but I think it will be important to be able to _skip_ them or focus specific tests._

Possibly more! Feel free to discuss!

(cc @JakeDawkins who has been thinking about this)

**Comment by Enrico2** [comment]
I can only assume there will be some test function in rust annotated with `#[test]` to run the suite, ignoring it is as easy as adding `#[ignore]`.

If we want this at an individual test _case_ level, I suspect we're going to end up having to add it to the .feature files and support it in the shim.

I'm fine with the all or nothing case because we can always locally remove the `#[ignore]` attribute until everything passes.

**Comment by Enrico2** [comment]
It'd be good to add some test cases to the .feature files that test mutations, currently there are none and we have no coverage for the mutation case. 

https://codecov.io/gh/apollographql/apollo-server/src/main/packages/apollo-gateway/src/buildQueryPlan.ts

**Comment by abernix** [comment]
Oooo, good looking out, @Enrico2! Opened https://github.com/apollographql/apollo-server/issues/4384 to track that.

**Comment by JakeDawkins** [comment]
@Enrico2 @abernix thinking about this, we actually _do_ have some existing mutation tests in the gateway's [existing integration test suite](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-gateway/src/__tests__/integration/mutations.test.ts#L207-L280).

Those integration tests test execution of the gateway and the data returned, but they also test query plan shape. We could pull out the query planning portion of those tests into the existing query plan tests if y'all think that'd be a good idea?

I think the one gotcha is we'd probably want to use multiple feature files, similar to that directory's structure now, so we'd maybe want to discuss putting all those tests in a compliance testing repo of some sort sooner rather than later

**Comment by Enrico2** [comment]
Can you explain why we'd want multiple feature files?

**Comment by JakeDawkins** [comment]
I think the general idea of a _feature_ file is to test a certain feature. So logically if we had a test suite to test the query planner, we'd probably want to organize the tests by what feature of query planning they were testing. 

That and finding and managing the suite of tests as they grow could become more a burden. Like, without knowing the name of a test, if you were looking to see if something specific was already tested against, it'd be much easier to find without searching a multi-thousand line file. 

If we were to move a bunch of tests from the integration suite to the query planning tests, I think it may be a good time to keep those tests organized similar to how they are now.

Like the test file for aliases, abstract types, etc in that `apollo-gateway/src/__tests__/integration` dir. Does that make sense @Enrico2?

**Comment by Enrico2** [comment]
I was considering "query planning" as a single feature, but I take your point about the feature file ending up huge. Ideally breaking them to logical parts of query planning sounds good.
Having said that, personally I'd rather we weren't blocked on #104 to have something _working_, but then again I don't really know how complex it's going to be to get #104 done right.

**Comment by JakeDawkins** [comment]
Yeah, that makes complete sense. I'll get the couple mutation tests we _do_ have moved into that file and PR'd in the next little bit! I'll also add a description to #106 and get it ready to review soon so you can use it :)
